[stdlib] Review and fix some problems with unsafe buffer and Range initialization

lorentey opened the pull request:
The goal here is to make sure there is no performance argument against using `UnsafeBufferPointer` instead of passing around naked pointer/count values, while also making sure obvious errors are caught in debug builds at least.

1. Make `U*BP.init(start:count:)` only check for invalid input in debug builds (by @karwa in #34747)
2. Add a missing debug mode check in `[Closed]Range.init(uncheckedBounds:)` (from #34951)
3. Add a missing debug mode check in `U*BP.init(rebasing:)` (followup to #34879 based on discussions therein)
4. Eliminate a signed overflow check in `U[M]BP.distance(from:to:)` (ditto)
5. Allow the compiler the assume a nonnegative result in `U[M]RBP.count` (bonus change)

Putting on my pedantic hat, the new `_debugPreconditions` are technically source-breaking changes. However, for point (2), `Range.init(uncheckedBounds:)` was documented to strictly require these conditions, and ranges that violate the invariant aren't likely to be usable. (Code that consumes ranges invariably assumes that `lowerBound <= upperBound`.)

As for point (3), `U*BP.init(rebasing:)` can only ever trap if someone explicitly invoked the `Slice(base:bounds:)` initializer with a bogus index range -- which is highly unlikely, since passing that to `.init(rebasing:)` is far more work than just calling `.init(start:count:)`.

ABI compatibility can only be an issue for binaries built in debug mode. (Which may be a significant problem.)

If compatibility concerns prevent us from adding (2) and (3) above, then we should remove these but keep the rest.

karwa committed:
Make UBP nil/count checks debug-mode only

Every other Unsafe(Mutable)BufferPointer precondition in this file is debug-mode only.

This information is already part of the documentation for this function, and it's really hard to get rid of these branches and traps otherwise.

karwa committed:
Make raw buffer nil/count checks debug-mode only

lorentey committed:
[stdlib] Encourage more inlining on String’s UTF-8 decoding fast path

lorentey committed:
[test] ArrayTraps: Skip trap test on negative UBP counts in release builds

lorentey committed:
[stdlib] Add a missing debug precondition to [Closed]Range.init(uncheckedBounds:)

Unchecked APIs must still perform checks in debug builds to ensure that invariants aren’t violated.
(I.e., these aren’t a license to perform invalid operations — they just let us get rid of the checks when we know for sure they are unnecessary.)

lorentey committed:
[stdlib] Add a missing debug mode check to U*BP.init(rebasing:)

`Slice` does not technically guarantee that its indices are valid in its base, so these initializers accidentally allowed the creation of obviously out-of-bounds buffers.

lorentey committed:
[stdlib] Eliminate an overflow check in U[M]BP.distance

lorentey committed:
[stdlib] U[M]RBP.count: Let the compiler assume that the result is nonnegative

lorentey commented:
Cc @karwa @Lukasa

lorentey commented:
Let's see how much havoc this will wreak in our benchmarks. 🙈 

@swift-ci benchmark

lorentey commented:
@swift-ci test

Lukasa commented:
> Let's see how much havoc this will wreak in our benchmarks. 🙈

I assume it will wreak _all_ the havoc. No benchmark will be safe.

Lukasa reviewed (approved):
Shiny, I love this patch.

lorentey commented:
@swift-ci benchmark

lorentey commented:
OK, the egregious code size regressions in #34951 are gone now. I'm assuming the time regressions are due to changes in inlining decisions; they don't seem to bad, surprisingly enough.

> ### Performance: -Osize
> 
> **Regression**                                    | **OLD** | **NEW** | **DELTA** | **RATIO**
> :---                                              | ---:  | ---:  | ---:    | ---:     
> SubstringFromLongString                           | 11    | 22    | +100.0% | **0.50x**
> FlattenListFlatMap                                | 3932  | 6740  | +71.4%  | **0.58x (?)**
> SubstringFromLongStringGeneric                    | 17    | 27    | +58.8%  | **0.63x**
> FlattenListLoop                                   | 1630  | 2509  | +53.9%  | **0.65x (?)**
> ObjectiveCBridgeStubDataAppend                    | 2760  | 3640  | +31.9%  | **0.76x**
> ArrayAppendSequence                               | 810   | 1040  | +28.4%  | **0.78x (?)**
> ObjectiveCBridgeStringHash                        | 110   | 140   | +27.3%  | **0.79x**
> Data.append.Sequence.64kB.Count0.RE               | 295   | 344   | +16.6%  | **0.86x (?)**
> Data.append.Sequence.64kB.Count0.RE.I             | 295   | 343   | +16.3%  | **0.86x**
> ObjectiveCBridgeStringGetASCIIContents            | 444   | 513   | +15.5%  | **0.87x (?)**

`SubstringFromLongString` is just calling `Substring(string)` with a long native string. The `Generic` variant does the same in a function that's generic over RangeReplaceableCollection. There is no reason that these should have become slower, so that's definitely worth a look. (Although I'm guessing the test is probably unduly emphasizing inlining decisions in a particular context that may not be a good indication of code performance in general.)

`FlattenListFlatMap` looks like an array creation & `append(contentsOf:)` microbenchmark. I'm surprised it isn't dominated by allocation overhead (do the intermediates get stack-allocated?), and it's difficult to see where the regressions may have come from. Again worth a look, if only to understand what's going on. It looks like this is in the same group as `FlattenListLoop` and `ArrayAppendSequence`; these are all benchmarking `Array.append()`/`Array.append(contentsOf:)`. Appending to an array got much faster in some cases in `-O`, so this smells like an optimizer instability. Not looking forward to diagnosing this.

The `ObjectiveCBridge*` regressions are mysterious to me as they don't seem to be dealing with ranges or buffer pointers at all. They are also not that large. I can revisit them once I know more about the above two groups.

lorentey commented:
macOS test failure is unrelated.

```
14:25:48 ******************** TEST 'Swift(iphonesimulator-i386) :: IRGen/async/run-switch-executor.swift' FAILED ********************
14:25:48 Child process terminated with signal 10: Bus error
```

karwa commented:
Fantastic, thanks for taking this on @lorentey. I wasn’t sure how to debug the benchmark regressions, so I was worried I wouldn’t be able to land the other patch on my own. I’ve heard that seemingly unrelated regressions can sometimes be down to code alignment, but I don’t know how to check that.

Anyway, I’m really happy to see these improvements.

lorentey committed:
[stdlib] Substring: Restore original Range paths

This eliminates a couple of _debugPreconditions which seem to change inlining enough to interfere with some benchmarks.

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci test

lorentey commented:
> ### Performance: -Osize
> 
> **Regression**                                  | **OLD** | **NEW** | **DELTA** | **RATIO**
> :---                                            | ---: | ---:  | ---:   | ---:     
> ObjectiveCBridgeStubDataAppend                  | 2500 | 3260  | +30.4% | **0.77x**
> ArrayAppendSequence                             | 700  | 910   | +30.0% | **0.77x (?)**
> ObjectiveCBridgeStringHash                      | 99   | 117   | +18.2% | **0.85x**
> Data.append.Sequence.64kB.Count0.RE.I           | 265  | 307   | +15.8% | **0.86x (?)**
> Data.append.Sequence.64kB.Count0.RE             | 265  | 306   | +15.5% | **0.87x**
> Data.init.Sequence.64kB.Count0.RE.I             | 269  | 310   | +15.2% | **0.87x (?)**
> Data.init.Sequence.64kB.Count0.RE               | 269  | 310   | +15.2% | **0.87x (?)**
> Data.init.Sequence.64kB.Count0.I                | 272  | 311   | +14.3% | **0.87x (?)**
> Data.init.Sequence.64kB.Count0                  | 273  | 311   | +13.9% | **0.88x (?)**
> Data.append.Sequence.64kB.Count0                | 274  | 312   | +13.9% | **0.88x (?)**
> Data.append.Sequence.64kB.Count0.I              | 274  | 312   | +13.9% | **0.88x (?)**
> Data.append.Sequence.809B.Count0.RE.I           | 370  | 420   | +13.5% | **0.88x (?)**
> Data.init.Sequence.2047B.Count0.I               | 493  | 554   | +12.4% | **0.89x**

That's better. `FlattenListFlatMap` still has a regression in `-O` mode though.

> ### Performance: -O
> 
> **Regression**                                   | **OLD** | **NEW** | **DELTA** | **RATIO**
> :---                                             | ---:  | ---:  | ---:    | ---:     
> FlattenListFlatMap                               | 5457  | 23581 | +332.1% | **0.23x**
> ObjectiveCBridgeStubDataAppend                   | 2420  | 3280  | +35.5%  | **0.74x**
> ObjectiveCBridgeStringHash                       | 99    | 117   | +18.2%  | **0.85x**
> Data.init.Sequence.64kB.Count0.RE.I              | 253   | 295   | +16.6%  | **0.86x**
> Data.init.Sequence.64kB.Count0                   | 251   | 292   | +16.3%  | **0.86x**
> Data.init.Sequence.64kB.Count0.RE                | 253   | 294   | +16.2%  | **0.86x**
> Data.init.Sequence.64kB.Count0.I                 | 252   | 292   | +15.9%  | **0.86x**

lorentey committed:
[stdlib] Use [Closed]Range.init(_uncheckedBounds:) in more places

When the stdlib says not to check, it’s a good idea to actually not have any checking, so that we leave existing code paths unchanged. (And because we do trust that the stdlib is responsible about such things.)

lorentey committed:
[stdlib] Don’t use unbound ranges (`foo[…]`) in String → Substring conversions

The UnboundRange → Range conversion path is complicated and it involves at least one unnecessary _precondition check for startIndex ..< endIndex.

lorentey commented:
The latest commits restore even more range conversion paths, and make some low-hanging simplifications.

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci benchmark

lorentey committed:
[stdlib] U[R]BP: Restore compile-time condition lost with removed precondition

lorentey commented:
Regarding that `FlattenListFlatMap` regression in `-O`: it looks like the intermediate four-element arrays below are optimized away completely on `main`, but they are stack-allocated on this branch.

```swift
func flattenFlatMap(_ input: [(Int, Int, Int, Int)]) -> [Int] {
  return input.flatMap { [$0.0, $0.1, $0.2, $0.3] }
}
```

This is somehow triggered by the compiler losing evidence about `count` being nonnegative by the change of `_precondition` to `_debugPrecondition` below.

```swift
  public init(
    @_nonEphemeral start: Unsafe${Mutable}Pointer<Element>?, count: Int
  ) {
    _debugPrecondition(
      count >= 0, "Unsafe${Mutable}BufferPointer with negative count")
    _debugPrecondition(
      count == 0 || start != nil,
      "Unsafe${Mutable}BufferPointer has a nil start and nonzero count")
    _position = start
    self.count = count
  }
```

Replacing the assignment with `self.count = _assumeNonNegative(count)` seems to let the optimizer get its mojo back.

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci test

lorentey commented:
Remaining regressions seem to be due to `_assumeNonnegative(count)` not being enough -- the `count >= 0` precondition evidently enables optimizations that this is currently missing. (FWIW, `_uncheckedUnsafeAssume` slows down `FlattenListFlatMap` by another factor of 2.) Or for all I know it could just be an inlining instability -- I have no idea what I'm doing. ;-) (The weird part is that the SIL/disassembly looks the same before/after; the difference may be hiding in one of the stdlib dylib entry points.)

The macOS test regression is another case of `utf8_decoding_fastpath.swift`:

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-main/swift/test/SILOptimizer/utf8_decoding_fastpath.swift:84:15: error: CHECK-DAG: expected string not found in input
// CHECK-DAG: function_ref {{.*}}_fromUTF8Repairing
              ^
<stdin>:690:84: note: scanning from here
sil @$s22utf8_decoding_fastpath022decodeStringUTF8ViewAsF0yS2S0fG0VF : $@convention(thin) (@guaranteed String.UTF8View) -> @owned String {
                                                                                   ^
<stdin>:698:7: note: possible intended match here
 %5 = function_ref @$sSS8UTF8ViewV32withContiguousStorageIfAvailableyxSgxSRys5UInt8VGKXEKlFSS_Tg50104$sSRy7ElementSTQzGSSs5Error_pIgyozo_ACSSsAD_pIegyrzo_SlRzs16_UnicodeEncodingR_8CodeUnitQy_AARtzr0_lTRSS8ab42V_s0C0O0G0OTG5033$sSS8decoding2asSSx_q_mtcst10_cd3R_8ef4Y58_7a26Rtzr0_lufcSSSRyAGGXEfU_SS8gH14V_s0C0O0H0OTg5Tf3nnpf_nTf1cn_n : $@convention(thin) (@guaranteed String.UTF8View) -> (@owned Optional<String>, @error Error) // user: %7
      ^
```

The tiny addition of the `_assumeNonnegative` call made the compiler decide not to inline either decoding closure this time, which is Just Great. This time I'm resolving it by adding an alternate UBP initializer, like with `Range.init(_uncheckedBounds:)`.

lorentey committed:
[stdlib] Add an unsafe U[M]BP initializer to work around some inliner test failures

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci benchmark

milseman commented on stdlib/public/core/UnsafeBufferPointer.swift.gyb:
```diff
@@ -387,6 +394,15 @@ extension Unsafe${Mutable}BufferPointer: ${Mutable}Collection, RandomAccessColle
+    _position = start
+    self.count = count
+  }
+
```
Should this also be `@inlinable`? I don't think AEIC implies that, but I don't recall the details.

What is `@_nonEphemeral`?

lorentey replied:
`@_aeic` does imply `@inlinable` -- indeed, it is a stronger variant of it. 

The `@_nonEphemeral` attribute indicates that the associated function escapes the pointer. The compiler emits a warning when a temporary pointer (such as the ones resulting from the implicit inout-to-pointer conversion feature) is passed for a parameter that has this attribute. It's a nice (if limited) safety feature for a common issue.

milseman reviewed (commented):
Overall I'm in favor of this change! I would like to see the optimizer be smarter about this code, what's the best way to follow up on that?

lorentey commented:
@milseman I have not looked at the implementation of the optimizer, but based the behavior exhibited here and similar cases (as well as second/third-hand info), it seems that one major practical source of performance instability is that the presence/absence of a `_debugPrecondition`/`assert` can affect inlining decisions in optimized builds. (I assume because they count towards some sort of a function size metric that is collected before constant folding.) This isn't right -- AFAICT these calls are ~guaranteed to get completely eliminated without a trace, so they shouldn't prevent any optimization opportunities.

If this diagnosis is correct (which is a big if -- recall that I know very little about how the optimizer works), then one piece of duck tape we could apply is to explicitly tell the compiler about this fact. I.e., we could introduce a semantic attribute that entirely removes these calls at a very early stage in optimized mode (or just changes any code size metric to not count their projection). This would be an extremely limited half-solution, but it may still be worth it, given how much we rely on these assertions. @atrick @aschwaighofer @eeckstein @gottesmm What do you think? As a Swift engineer, I expect to be able to add `assert` calls with wild, reckless abandon, without worrying about their potential effect on production builds.

lorentey commented:
Yep, https://github.com/apple/swift/pull/35065 demonstrates that `_debugPrecondition` has a measurable effect on optimized code. I think it should not.

```swift
@usableFromInline @_transparent
internal func _debugPrecondition(
  _ condition: @autoclosure () -> Bool, _ message: StaticString = StaticString(),
  file: StaticString = #file, line: UInt = #line
) {
  // Only check in debug mode.
  if _slowPath(_isDebugAssertConfiguration()) {
    if !_fastPath(condition()) {
      _fatalErrorMessage("Fatal error", message, file: file, line: line,
        flags: _fatalErrorFlags())
    }
  }
```

lorentey committed:
[stdlib] Document sources of code bloat

lorentey commented:
@swift-ci test

eeckstein commented:
@lorentey Unfortunately the situation is more complicated.

We have to distinguish between debug/release build of the stdlib itself and debug/release build of client code.
It's _not_ the case the uses of `assert`/`_debugPrecondition` in the stdlib are removed in release builds (of the client program). It cannot, because the same release-built stdlib is used to for debug _and_ release builds of client programs.

`assert`/`_debugPrecondition` is only removed if
* used in the client code (and the client is built in release)
* a stdlib function is specialized in the client module. But this is not something one can rely on.

Now, I'm not sure what you are asking for. In case you are looking for an "assert" which is removed in the release build of the stdlib: it's `_internalInvariant` (as you probably know).

Or do you mean that specialized stdlib functions should result in the same code regardless if an `assert`/`_debugPrecondition` is in the function's implementation? I'm not sure why this should be a goal, because we cannot guarantee this in general, i.e. in the case a function is not specialized and called in the stdlib dylib.

All this is quite confusing and it always takes me a while until I fully understand it.

lorentey commented:
That makes sense! However, at least some of the issues I've been fighting was about compiler behavior for code that isn't calling into the stdlib dylib -- e.g., the `utf8_decoding_fastpath.swift` test that was the most troublesome in #34747 is looking entirely at code that's emitted into the client module. (Then again, for all I know, changing _precondition to _debugPrecondition may have somehow caused less code getting inlined. E.g., eliminating a `>= 0` check up front may have induced more code emitted to handle the `< 0` case later.)

I also don't recall seeing any calls into the swift parts of libswiftCore in the FlattenListFlatMap hot loop that I needed to investigate above -- and that one is listed as (weakly) impacted in #35065.

I can easily believe how the AnySequence/AnyCollection benchmarks (that got a 2x speedup in #35065) may call into libswiftCore though -- their code is fully inlinable but I don't expect their boxes would get specialized. The 2x factor seems a bit harsh; we should probably track down the exact cause and try to fix it somehow.

lorentey commented:
> I also don't recall seeing any calls into the swift parts of libswiftCore in the FlattenListFlatMap hot loop that I needed to investigate above -- and that one is listed as (weakly) impacted in #35065.

Hm, I may be misremembering; I [explicitly speculated](https://github.com/apple/swift/pull/34961#issuecomment-742213241) on dylib changes triggering the issue above. I'll need to double check.

lorentey commented:
Yeah, FlattenListFlatMap only calls runtime entry points and a specialized version of `_ArrayBuffer<Int
>._consumeAndCreateNew` that is within the same module. The generated code is, as far as I can tell, exactly the same before/after removing `_debugPrecondition`'s body.

The FlattenListFlatMap benchmark results in #35065 seem to be entirely due to benchmark flakiness -- locally measured numbers seem to be roughly similar.

Phew! 😌

lorentey commented:
`SequenceAlgosAnySequence` seems to be a good laboratory sample for the 2x speedups in #35065. It is calling this function with `s == AnySequence(0 ..< 100_000)`:

```swift
func benchmarkSequenceAlgos<S: Sequence>(s: S, n: Int) where S.Element == Int {
  CheckResults(s.reduce(0, &+) == (n*(n-1))/2)
  let mn = s.min()
  let mx = s.max()
  CheckResults(mn == 0 && mx == n-1)
  CheckResults(s.starts(with: s))
}
```

This gets correctly specialized for `AnySequence<Int>`, and it essentially turns into a benchmark for `_AnyIteratorBoxBase.next()`, which lives in libswiftCore.dylib. Unfortunately that calls `Range.subscript` that includes a _debugPrecondition that regretfully it cannot be eliminated within the stdlib itself. (ಥ﹏ಥ)

I think the large speedups in #35065 are all because of this wrinkle. (I don't think they are cause for alarm, though. Who uses collection existentials on ranges in practice?) 

I have not looked into the `ObjectiveCBridge*` regressions in the same PR; I expect they might be depending on `count >= 0` invariants that the compiler cannot prove once the preconditions are gone, like with the `utf8_decoding_fastpath` test.

This leaves the `FlattenListLoop`, `InsertCharacterEndIndex` and `NSStringConversion.MutableCopy.Rebridge.LongUTF8` regressions which are (like `FlattenListFlatMap`) marked as unreliable in the benchmark report, and I don't see why we shouldn't blame them on flakiness.

So I think I have convinced myself everything is "fine" -- `_debugPrecondition` works as designed. 

Unfortunately we'll likely need to stop relying on the collection conformances of `[Closed]Range` and `Unsafe[Mutable][Raw]BufferPointer` within the internals of the stdlib -- for example, having bounds checks has serious consequences for any generic algorithm that calls `withContiguous[Mutable]StorageIfAvailable` within the stdlib as an optimization. (As, say, `MutableCollection.partition(by:)` does in its random-access implementation.)

milseman commented:
Precondition/debugPrecondition/internalInvariant is mentioned in https://github.com/apple/swift/blob/main/docs/StandardLibraryProgrammersManual.md#_precondition-_debugprecondition-and-_internalinvariant.
